File: backend/app/schemas/study.py
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field

from backend.app.models.entities import TaskStatus


class CourseCreate(BaseModel):
    name: str = Field(..., min_length=2, max_length=120)
    description: str = Field(default="", max_length=500)


class CourseResponse(BaseModel):
    id: str
    name: str
    description: str
    created_at: datetime


class StudyTaskCreate(BaseModel):
    title: str = Field(..., min_length=2, max_length=160)
    course_id: str | None = None
    prompt: str = Field(default="", max_length=1000)
    material_text: str = Field(..., min_length=30)
    source_type: Literal["prompt", "text", "pdf"] = "text"
    source_name: str | None = None


class StudyTaskResponse(BaseModel):
    id: str
    title: str
    course_id: str | None
    prompt: str
    source_type: str
    source_name: str | None
    status: TaskStatus
    progress: int
    result: dict[str, Any] | None
    error: str | None
    created_at: datetime
    updated_at: datetime


class HealthResponse(BaseModel):
    status: str
    service: str

File: frontend/src/types/study.ts
export type Course = {
  id: string;
  name: string;
  description: string;
  created_at: string;
};

export type QuizQuestion = {
  question: string;
  answer: string;
  topic: string;
  options?: string[];
  correct_option?: string;
};

export type StudyPlanDay = {
  day: number;
  focus: string;
  tasks: string[];
};

export type StudyResult = {
  title: string;
  important_topics: string[];
  summary: string[];
  quiz: QuizQuestion[];
  study_plan: StudyPlanDay[];
  generation?: {
    tier: "offline" | "free" | "codex";
    provider: string;
    model: string;
    fallback_reason?: string | null;
  };
};

export type StudyTask = {
  id: string;
  title: string;
  course_id: string | null;
  prompt: string;
  source_type: string;
  source_name: string | null;
  status: "pending" | "processing" | "completed" | "failed";
  progress: number;
  result: StudyResult | null;
  error: string | null;
  created_at: string;
  updated_at: string;
};

export type StudyTaskPayload = {
  title: string;
  course_id?: string;
  prompt: string;
  material_text: string;
  source_type: "prompt" | "text" | "pdf";
  source_name?: string;
};
